File: android/src/main/java/com/guichaguri/trackplayer/service/models/TrackMetadata.java
package com.guichaguri.trackplayer.service.models;

import android.content.Context;
import android.net.Uri;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.RatingCompat;
import com.facebook.react.bridge.ReadableMap;
import com.guichaguri.trackplayer.service.Utils;

import static android.support.v4.media.MediaMetadataCompat.*;

public abstract class TrackMetadata {
    public Uri artwork;

    public String title;
    public String artist;
    public String album;
    public String date;
    public String genre;
    public long duration;

    public RatingCompat rating;

    public void setMetadata(Context context, ReadableMap data, int ratingType) {
        artwork = Utils.getUri(context, data, "artwork");

        title = data.getString("title");
        artist = data.getString("artist");
        album = data.getString("album");
        date = data.getString("date");
        genre = data.getString("genre");
        duration = Utils.toMillis(Utils.getDouble(data, "duration", 0));

        rating = Utils.getRating(data, "rating", ratingType);
    }

    public MediaMetadataCompat.Builder toMediaMetadata() {
        MediaMetadataCompat.Builder builder = new MediaMetadataCompat.Builder();

        builder.putString(METADATA_KEY_TITLE, title);
        builder.putString(METADATA_KEY_ARTIST, artist);
        builder.putString(METADATA_KEY_ALBUM, album);
        builder.putString(METADATA_KEY_DATE, date);
        builder.putString(METADATA_KEY_GENRE, genre);

        if (duration > 0) {
            builder.putLong(METADATA_KEY_DURATION, duration);
        }

        if (artwork != null) {
            builder.putString(METADATA_KEY_ART_URI, artwork.toString());
        }

        if (rating != null) {
            builder.putRating(METADATA_KEY_RATING, rating);
        }

        return builder;
    }

}

File: android/src/main/java/com/doublesymmetry/trackplayer/model/TrackMetadata.kt
package com.doublesymmetry.trackplayer.model

import android.content.Context
import android.net.Uri
import android.os.Bundle
import androidx.media3.common.Rating
import com.doublesymmetry.trackplayer.extensions.NumberExt.Companion.toMilliseconds
import com.doublesymmetry.trackplayer.utils.BundleUtils

abstract class TrackMetadata {
    var artwork: Uri? = null
    var title: String? = null
    var artist: String? = null
    var album: String? = null
    var date: String? = null
    var genre: String? = null
    var duration: Long? = null
    var rating: Rating? = null
    var mediaId: String? = null

    open fun setMetadata(context: Context, bundle: Bundle?, ratingType: Int) {
        // Special handling for artwork:
        // - If undefined/null in bundle - keep existing artwork
        // - If empty string - explicitly clear artwork
        // - If valid URI - update artwork
        if (bundle?.containsKey("artwork") == true) {
            val artworkStr = bundle.getString("artwork")
            artwork = if (artworkStr?.isEmpty() == true) {
                null // Clear artwork for empty string
            } else {
                BundleUtils.getUri(context, bundle, "artwork") ?: artwork // Keep existing if new is invalid
            }
        }

        // For all other fields, only update if provided in bundle
        bundle?.getString("title")?.let { title = it }
        bundle?.getString("artist")?.let { artist = it }
        bundle?.getString("album")?.let { album = it }
        bundle?.getString("date")?.let { date = it }
        bundle?.getString("genre")?.let { genre = it }
        bundle?.getString("mediaId")?.let { mediaId = it }

        // Update duration if provided
        if (bundle?.containsKey("duration") == true) {
            duration = bundle.getDouble("duration").toMilliseconds()
        }

        // Update rating if provided
        if (bundle?.containsKey("rating") == true) {
            rating = BundleUtils.getRating(bundle, "rating", ratingType)
        }
    }
}
